runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/state/WindmillMapViaMultimap.java

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.state;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill;
+import org.apache.beam.sdk.state.ReadableState;
+import org.apache.beam.sdk.state.ReadableStates;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Supplier;
+
+public class WindmillMapViaMultimap<KeyT, ValueT> extends AbstractWindmillMap<KeyT, ValueT> {
+  final WindmillMultimap<KeyT, ValueT> multimap;
+
+  WindmillMapViaMultimap(WindmillMultimap<KeyT, ValueT> multimap) {
+    this.multimap = multimap;
+  }
+
+  @Override
+  protected Windmill.WorkItemCommitRequest persistDirectly(WindmillStateCache.ForKeyAndFamily cache)
+      throws IOException {
+    return multimap.persistDirectly(cache);
+  }
+
+  @Override
+  void initializeForWorkItem(
+      WindmillStateReader reader, Supplier<Closeable> scopedReadStateSupplier) {
+    super.cleanupAfterWorkItem();
+    multimap.initializeForWorkItem(reader, scopedReadStateSupplier);
+  }
+
+  @Override
+  void cleanupAfterWorkItem() {
+    super.cleanupAfterWorkItem();
+    multimap.cleanupAfterWorkItem();
+  }
+
+  @Override
+  public void put(KeyT key, ValueT value) {
+    multimap.remove(key);
+    multimap.put(key, value);
+  }
+
+  @Override
+  public ReadableState<ValueT> computeIfAbsent(
+      KeyT key, Function<? super KeyT, ? extends ValueT> mappingFunction) {
+    // Note that computeIfAbsent comments indicate that the read is lazy but this matches the
+    // existing eager
+    // behavior of WindmillMap.
+    Iterable<ValueT> existingValues = multimap.get(key).read();
+    if (Iterables.isEmpty(existingValues)) {
+      ValueT inserted = mappingFunction.apply(key);
+      multimap.put(key, inserted);
+      return ReadableStates.immediate(inserted);
+    } else {
+      return ReadableStates.immediate(Iterables.getOnlyElement(existingValues));
+    }
+  }
+
+  @Override
+  public void remove(KeyT key) {
+    multimap.remove(key);
+  }
+
+  private static class SingleValueIterableAdaptor<T> implements ReadableState<T> {
+    final ReadableState<Iterable<T>> wrapped;
+    final @Nullable T defaultValue;
+
+    SingleValueIterableAdaptor(ReadableState<Iterable<T>> wrapped, @Nullable T defaultValue) {
+      this.wrapped = wrapped;
+      this.defaultValue = defaultValue;
+    }
+
+    @Override
+    public T read() {
+      Iterator<T> iterator = wrapped.read().iterator();
+      if (iterator.hasNext()) {
```

Comment:
If `!hasNext()'?

Reply:
Looks like the tests could use some improvement. I updated the validates runner test to cover a little more but I will also going to add unit tests to WindmillStateInternalsTest. I'll let you know when it's ready for another look.

Reply:
Added some unit tests